python/Ganga/GPIDev/Base/Proxy.py

```diff
@@ -811,51 +820,10 @@ def _init(self, *args, **kwds):
         for k in kwds:
             if stripProxy(self)._schema.hasAttribute(k):
                 this_arg = stripProxy(kwds[k])
-                if hasattr(this_arg, '_auto__init__'):
-                    this_arg._auto__init__()
-
-                ## Copying this from the __set__ method in the Proxy descriptor
-
-                if k == 'application':
-                    ProxyDataDescriptor.__app_set__(self, this_arg)
-                if k == 'is_prepared':
-                    ProxyDataDescriptor.__prep_set__(self, this_arg)
-
-
                 raw_self = stripProxy(self)
-
-                if isinstance(this_arg, str):
-                    this_arg = stripProxy(runtimeEvalString(raw_self, k, this_arg))
-                    if hasattr(this_arg, '_auto__init__'):
-                        this_arg._auto__init__()
-
-                if isinstance(this_arg, str):
-                    raw_self.setSchemaAttribute(k, raw_self._attribute_filter__set__(k, this_arg))
-                    continue
-                else:
-                    item = pluginclass._schema.getItem(k)
-
-                    # unwrap proxy
-                    if item.isA(ComponentItem):
-                        from .Filters import allComponentFilters
-                        cfilter = allComponentFilters[item['category']]
-                        stripper = lambda v: stripComponentObject(v, cfilter, item)
-                    else:
-                        stripper = None
-
-                    if item['sequence']:
-                        this_arg = ProxyDataDescriptor.__sequence_set__(stripper, raw_self, this_arg, k)
-                    else:
-                        if stripper is not None:
-                            this_arg = stripper(this_arg)
-                    # apply attribute filter to component items
-                    if item.isA(ComponentItem):
-                        this_arg = ProxyDataDescriptor._stripAttribute(raw_self, this_arg, k)
-
-                    if hasattr(this_arg, '_auto__init__'):
-                        this_arg._auto__init__()
-
-                    raw_self.setSchemaAttribute(k, raw_self._attribute_filter__set__(k, this_arg))
+                # This calls the same logic when assigning a named attribute as when we're assigning it to the object
+                # There is logic here which we 'could' duplicate but it is over 100 lines of code which then is duplicating funtionality written elsewhere
+                ProxyDataDescriptor(k, False).__set__(raw_self, raw_self._attribute_filter__set__(k, this_arg))
```

Comment:
There's no such thing as read-only schema attributes, only read-only objects.

It looks like the real problem is in `export`ing and `load`ing jobs in the GPI. For example, a job which has been creaded and submitted and is then `export(j)`ed will make a file like:
```python
Job (
   comment = '',
   parallel_submit = False,
   do_auto_resubmit = False,
   name = '',
   inputfiles =[],
   backend =   Local (
     nice = 0
   ),
   application =   Executable (
     exe = 'echo',
     hash = 'cfa689697ba25ae38f7a47b232ea3e7c',
     env = [],
     args = ['Hello World'],
     is_prepared = ShareDir(name='conf-47de7c78-e402-4c59-af7f-29138635fcfa',subdir='.')
   ),
   inputdata =None,
   outputfiles =[],
   postprocessors =   MultiPostProcessor (
   ),
   info =   JobInfo (
     monitor =None
   ),
   splitter =None,
   inputsandbox =[]
 )
```

The problem come when Python is parsing the
```python
Executable (
  exe = 'echo',
  hash = 'cfa689697ba25ae38f7a47b232ea3e7c',
  env = [],
  args = ['Hello World'],
  is_prepared = ShareDir(name='conf-47de7c78-e402-4c59-af7f-29138635fcfa',subdir='.')
)
```
part. Since the order in which the arguments are processed is not fixed, the `is_prepared` attribute is being assigned before e.g. the `exe` attribute. This means that Ganga raises an error saying "This object is in the prepared state so you cannot change it".

The underlying problem here is that we want to ignore the read-only state of an object while it is being constructed and only start enforcing it once its state is fully defined. I'm working on a patch to this PR which will fix this is a slightly different way.